From 8f3dc384c6cf5dfc25025d5caae2a302f4ff540d Mon Sep 17 00:00:00 2001
From: Subramanian Srinivasan <subrsrin@codeaurora.org>
Date: Wed, 29 Apr 2020 13:43:51 -0700
Subject: [PATCH] Native stack access for certain GATT Notifications

Route GATT notifications received for certain GATT
characteristics from BT stack through native layer
socket to intended process.
This reduces latency in processing GATT notifications
for certain use cases.

Change-Id: I13b32f3b6a2646602303a17fa7d3445b521594c7
---
 bta/gatt/bta_gattc_act.cc | 169 +++++++++++++++++++++++++++++++++++++++++++++-
 bta/gatt/bta_gattc_int.h  |   9 +++
 stack/include/gattdefs.h  |   5 ++
 3 files changed, 182 insertions(+), 1 deletion(-)

diff --git a/bta/gatt/bta_gattc_act.cc b/bta/gatt/bta_gattc_act.cc
index d138ca8..2887466 100644
--- a/bta/gatt/bta_gattc_act.cc
+++ b/bta/gatt/bta_gattc_act.cc
@@ -45,6 +45,10 @@
 #include "bta_hh_int.h"
 #endif
 
+#include "osi/include/osi.h"
+#include "osi/include/socket_utils/sockets.h"
+#include "osi/include/properties.h"
+
 using base::StringPrintf;
 using bluetooth::Uuid;
 
@@ -113,11 +117,22 @@ void bta_gattc_reset_discover_st(tBTA_GATTC_SERV* p_srcb, tGATT_STATUS status);
 /** Enables GATTC module */
 static void bta_gattc_enable() {
   VLOG(1) << __func__;
+  char native_access_notif_prop[PROPERTY_VALUE_MAX] = "false";
 
   if (bta_gattc_cb.state == BTA_GATTC_STATE_DISABLED) {
     /* initialize control block */
     bta_gattc_cb = tBTA_GATTC_CB();
     bta_gattc_cb.state = BTA_GATTC_STATE_ENABLED;
+    bta_gattc_cb.is_gatt_skt_connected = false;
+    bta_gattc_cb.gatt_skt_fd = -1;
+    bta_gattc_cb.native_access_uuid_list = {Uuid::FromString(GATT_UUID_ACCEL_GYRO_STR),
+        Uuid::FromString(GATT_UUID_MAG_DATA_STR), Uuid::FromString(GATT_UUID_PRESSURE_SENSOR_STR)};
+
+    property_get(
+        "persist.vendor.btstack.enable.native_access_notification", native_access_notif_prop, "false");
+    if(!strcmp(native_access_notif_prop, "true")) {
+      bta_gattc_cb.native_access_notif_enabled = true;
+    }
   } else {
     VLOG(1) << "GATTC is already enabled";
   }
@@ -155,6 +170,12 @@ void bta_gattc_disable() {
     bta_gattc_cb = tBTA_GATTC_CB();
     bta_gattc_cb.state = BTA_GATTC_STATE_DISABLED;
   }
+
+  close(bta_gattc_cb.gatt_skt_fd);
+  bta_gattc_cb.is_gatt_skt_connected = false;
+  bta_gattc_cb.gatt_skt_fd = -1;
+  bta_gattc_cb.native_access_uuid_list.clear();
+  bta_gattc_cb.native_access_notif_enabled = false;
 }
 
 /** start an application interface */
@@ -1064,8 +1085,13 @@ static void bta_gattc_conn_cback(tGATT_IF gattc_if, const RawAddress& bdaddr,
 
   if (connected)
     btif_debug_conn_state(bdaddr, BTIF_DEBUG_CONNECTED, GATT_CONN_UNKNOWN);
-  else
+  else {
     btif_debug_conn_state(bdaddr, BTIF_DEBUG_DISCONNECTED, reason);
+    //close native socket during disconnection.
+    close(bta_gattc_cb.gatt_skt_fd);
+    bta_gattc_cb.is_gatt_skt_connected = false;
+    bta_gattc_cb.gatt_skt_fd = -1;
+  }
 
   tBTA_GATTC_DATA* p_buf =
       (tBTA_GATTC_DATA*)osi_calloc(sizeof(tBTA_GATTC_DATA));
@@ -1205,6 +1231,128 @@ bool bta_gattc_process_srvc_chg_ind(uint16_t conn_id, tBTA_GATTC_RCB* p_clrcb,
   return true;
 }
 
+/** build native access GATT notification data */
+uint8_t* bta_gattc_build_skt_notification_data(Uuid char_uuid,
+                                               tGATT_CL_COMPLETE* p_data,
+                                               size_t* total_len) {
+  size_t data_hdr_len = 4;
+  uint8_t char_uuid_type = 0x01;
+  uint8_t notif_data_type = 0x02;
+  size_t len_char_uuid = char_uuid.GetShortestRepresentationSize();
+  size_t len_data = p_data->att_value.len;
+  uint8_t* p_uuid = new uint8_t[len_char_uuid];
+  *total_len = len_char_uuid + len_data + data_hdr_len;
+  uint8_t* p = new uint8_t[*total_len];
+  uint8_t* pp = p;
+  uint8_t* p_tmp = pp;
+
+  sscanf(char_uuid.ToString().c_str(),
+         "%02hhx%02hhx%02hhx%02hhx-%02hhx%02hhx-%02hhx%02hhx"
+         "-%02hhx%02hhx-%02hhx%02hhx%02hhx%02hhx%02hhx%02hhx",
+         &p_uuid[0], &p_uuid[1], &p_uuid[2], &p_uuid[3], &p_uuid[4], &p_uuid[5],
+         &p_uuid[6], &p_uuid[7], &p_uuid[8], &p_uuid[9], &p_uuid[10], &p_uuid[11],
+         &p_uuid[12], &p_uuid[13], &p_uuid[14], &p_uuid[15]);
+
+  //LTV for char uuid
+  UINT8_TO_STREAM(pp, (len_char_uuid+1));
+  UINT8_TO_STREAM(pp, char_uuid_type);
+  ARRAY_TO_STREAM(pp, p_uuid, (int)len_char_uuid);
+
+  //LTV for notification data
+  UINT8_TO_STREAM(pp, (len_data+1));
+  UINT8_TO_STREAM(pp, notif_data_type);
+  ARRAY_TO_STREAM(pp, p_data->att_value.value, (int)len_data);
+
+  VLOG(1) << __func__ << " Socket data bytes: ";
+  for(uint8_t k=0; k<(data_hdr_len + len_char_uuid + len_data); k++) {
+    VLOG(1) << __func__ << " value: " << +p_tmp[k];
+  }
+
+  return p;
+}
+
+/** write native access GATT notification data to socket */
+bool bta_gattc_write_to_socket(tGATT_CL_COMPLETE* p_data, Uuid char_uuid) {
+  size_t count = 0, total_len = 0;
+  int sent = 0;
+  uint8_t* p_skt_data;
+
+  VLOG(1) << __func__;
+  if (bta_gattc_cb.gatt_skt_fd < 0) {
+    bta_gattc_cb.gatt_skt_fd = socket(AF_LOCAL, SOCK_SEQPACKET, 0);
+
+    if (bta_gattc_cb.gatt_skt_fd < 0) {
+      VLOG(1) << __func__ << " failed to create socket";
+      bta_gattc_cb.is_gatt_skt_connected = false;
+      close(bta_gattc_cb.gatt_skt_fd);
+      bta_gattc_cb.gatt_skt_fd = -1;
+      return false;
+    }
+  }
+
+  if (!bta_gattc_cb.is_gatt_skt_connected) {
+    if (osi_socket_local_client_connect(
+        bta_gattc_cb.gatt_skt_fd, BTA_GATTC_NATIVE_ACCESS_SOCKET, ANDROID_SOCKET_NAMESPACE_FILESYSTEM, SOCK_SEQPACKET) < 0) {
+      VLOG(1) << __func__ << " failed to connect: error: " << +strerror(errno);
+      close(bta_gattc_cb.gatt_skt_fd);
+      bta_gattc_cb.gatt_skt_fd = -1;
+      bta_gattc_cb.is_gatt_skt_connected = false;
+      return false;
+    }
+    else {
+      bta_gattc_cb.is_gatt_skt_connected = true;
+    }
+  }
+
+  p_skt_data = bta_gattc_build_skt_notification_data(char_uuid, p_data, &total_len);
+
+  //write to socket
+  while (count < total_len) {
+    OSI_NO_INTR(sent = send(bta_gattc_cb.gatt_skt_fd, p_skt_data, total_len - count, MSG_NOSIGNAL | MSG_DONTWAIT));
+    if (sent == -1) {
+      if (errno != EAGAIN && errno != EWOULDBLOCK) {
+        VLOG(1) << __func__ << "write failed with error:" << +strerror(errno);
+        bta_gattc_cb.is_gatt_skt_connected = false;
+        close(bta_gattc_cb.gatt_skt_fd);
+        bta_gattc_cb.gatt_skt_fd = -1;
+        return false;
+      }
+    }
+    count += sent;
+    p_skt_data = (uint8_t*)p_skt_data + sent;
+  }
+
+  return true;
+}
+
+/** process native access notification */
+void bta_gattc_proc_native_access_notification(tBTA_GATTC_SERV* p_srcb,
+                                    tGATT_CL_COMPLETE* p_data) {
+  VLOG(1) << __func__
+          << StringPrintf(": p_data->att_value.handle=%d",
+          p_data->att_value.handle);
+
+  std::vector<Uuid> char_uuid_list = bta_gattc_cb.native_access_uuid_list;
+  std::vector<Uuid>::iterator it;
+
+  uint16_t handle = p_data->att_value.handle;
+  const gatt::Characteristic* p_char = bta_gattc_get_characteristic_srcb(p_srcb, handle);
+  bool ret = false;
+  if(p_char) {
+    it = std::find (char_uuid_list.begin(), char_uuid_list.end(), p_char->uuid);
+    if (it != char_uuid_list.end()) {
+      VLOG(1) << __func__ << " GATT characteristic receiving notifications";
+      ret = bta_gattc_write_to_socket(p_data, p_char->uuid);
+      if(ret) {
+        VLOG(1) << __func__ << " GATT notification data write to socket successful";
+      }
+      else {
+        VLOG(1) << __func__ << " GATT notification data write to socket unsuccessful";
+      }
+    }
+  }
+}
+
 /** process all non-service change indication/notification */
 void bta_gattc_proc_other_indication(tBTA_GATTC_CLCB* p_clcb, uint8_t op,
                                      tGATT_CL_COMPLETE* p_data,
@@ -1214,6 +1362,25 @@ void bta_gattc_proc_other_indication(tBTA_GATTC_CLCB* p_clcb, uint8_t op,
                  ": check p_data->att_value.handle=%d p_data->handle=%d",
                  p_data->att_value.handle, p_data->handle);
   VLOG(1) << "is_notify" << p_notify->is_notify;
+  uint16_t conn_id = 0;
+  RawAddress remote_bda;
+  tGATT_IF gatt_if;
+  tBTA_TRANSPORT transport;
+
+  if(bta_gattc_cb.native_access_notif_enabled) {
+    //check for native access notification
+    if(op == GATTC_OPTYPE_NOTIFICATION) {
+      if (p_clcb) {
+        conn_id = p_clcb->bta_conn_id;
+        if (!GATT_GetConnectionInfor(conn_id, &gatt_if, remote_bda, &transport)) {
+          LOG(ERROR) << __func__ << ": notification for unknown app";
+          return;
+        }
+        tBTA_GATTC_SERV* p_srcb = bta_gattc_find_srcb(remote_bda);
+        bta_gattc_proc_native_access_notification(p_srcb, p_data);
+      }
+    }
+  }
 
   p_notify->is_notify = (op == GATTC_OPTYPE_INDICATION) ? false : true;
   p_notify->len = p_data->att_value.len;
diff --git a/bta/gatt/bta_gattc_int.h b/bta/gatt/bta_gattc_int.h
index 06e0913..d8da75e 100644
--- a/bta/gatt/bta_gattc_int.h
+++ b/bta/gatt/bta_gattc_int.h
@@ -36,6 +36,8 @@
 #include <base/logging.h>
 #include <base/strings/stringprintf.h>
 
+using bluetooth::Uuid;
+
 /*****************************************************************************
  *  Constants and data types
  ****************************************************************************/
@@ -84,6 +86,8 @@ typedef uint16_t tBTA_GATTC_INT_EVT;
 
 #define BTA_GATTC_WRITE_PREPARE GATT_WRITE_PREPARE
 
+#define BTA_GATTC_NATIVE_ACCESS_SOCKET "/dev/socket/qvrservice_controller"
+
 /* internal strucutre for GATTC register API  */
 typedef struct {
   BT_HDR hdr;
@@ -298,6 +302,11 @@ typedef struct {
 
   tBTA_GATTC_CLCB clcb[BTA_GATTC_CLCB_MAX];
   tBTA_GATTC_SERV known_server[BTA_GATTC_KNOWN_SR_MAX];
+
+  int gatt_skt_fd;
+  bool is_gatt_skt_connected;
+  std::vector<Uuid> native_access_uuid_list;
+  bool native_access_notif_enabled;
 } tBTA_GATTC_CB;
 
 /*****************************************************************************
diff --git a/stack/include/gattdefs.h b/stack/include/gattdefs.h
index 9dc013c..186aeab 100644
--- a/stack/include/gattdefs.h
+++ b/stack/include/gattdefs.h
@@ -131,4 +131,9 @@
 #define GATT_UUID_SCAN_INT_WINDOW 0x2A4F
 #define GATT_UUID_SCAN_REFRESH 0x2A31
 
+/* Native access notification characteristics */
+#define GATT_UUID_ACCEL_GYRO_STR       "00000001-2000-1000-8000-cec278b6b50a"
+#define GATT_UUID_MAG_DATA_STR         "00000002-2000-1000-8000-cec278b6b50a"
+#define GATT_UUID_PRESSURE_SENSOR_STR  "00000003-2000-1000-8000-cec278b6b50a"
+
 #endif
-- 
2.7.4

